release.sh

```diff
@@ -313,64 +314,109 @@ function build_from_source() {
 }
 
 function get_images_in_yamls() {
-  rm -rf imagerefs.txt
+  rm -rf "$IMAGES_REFS_FILE"
   echo "Assembling a list of image refences to sign"
-  for file in $@; do
+  for file in "$@"; do
     [[ "${file##*.}" != "yaml" ]] && continue
     echo "Inspecting ${file}"
-    for image in $(grep -oh "\S*${KO_DOCKER_REPO}\S*" "${file}"); do
-      echo $image >> imagerefs.txt
-    done
+    while read -r image; do
+      echo "$image" >> "$IMAGES_REFS_FILE"
+    done < <(grep -oh "\S*${KO_DOCKER_REPO}\S*" "${file}")
+  done
+  if [[ -f "$IMAGES_REFS_FILE" ]]; then
+    sort -uo "$IMAGES_REFS_FILE" "$IMAGES_REFS_FILE" # Remove duplicate entries
+  fi
+}
+
+function find_checksums_file() {
+  for file in "$@"; do
+    if [[ "${file}" == *"checksums.txt" ]]; then
+      echo "${file}"
+      return 0
+    fi
   done
-  sort -uo imagerefs.txt imagerefs.txt # Remove duplicate entries
+  warning "cannot find checksums file"
 }
 
 # Build a release from source.
 function sign_release() {
-  get_images_in_yamls "${ARTIFACTS_TO_PUBLISH}"
   if (( ! IS_PROW )); then # This function can't be run by devs on their laptops
     return 0
   fi
+  get_images_in_yamls "${ARTIFACTS_TO_PUBLISH}"
+  local checksums_file
+  checksums_file="$(find_checksums_file "${ARTIFACTS_TO_PUBLISH}")"
+
+  if ! [[ -f "${checksums_file}" ]]; then
+    echo '>> No checksums file found, generating one'
+    checksums_file="$(mktemp -d)/checksums.txt"
+    for file in ${ARTIFACTS_TO_PUBLISH}; do
+      pushd "$(dirname "$file")" >/dev/null
+      sha256sum "$(basename "$file")" >> "${checksums_file}"
+      popd >/dev/null
+    done
+    ARTIFACTS_TO_PUBLISH="${ARTIFACTS_TO_PUBLISH} ${checksums_file}"
+  fi
 
   # Notarizing mac binaries needs to be done before cosign as it changes the checksum values
   # of the darwin binaries
  if [ -n "${APPLE_CODESIGN_KEY}" ] && [ -n "${APPLE_CODESIGN_PASSWORD_FILE}" ] && [ -n "${APPLE_NOTARY_API_KEY}" ]; then
     banner "Notarizing macOS Binaries for the release"
-    FILES=$(find -- * -type f -name "*darwin*")
-    for file in $FILES; do
-      rcodesign sign "${file}" --p12-file="${APPLE_CODESIGN_KEY}" \
-        --code-signature-flags=runtime \
-        --p12-password-file="${APPLE_CODESIGN_PASSWORD_FILE}"
-    done
-    zip files.zip ${FILES}
-    rcodesign notary-submit files.zip --api-key-path="${APPLE_NOTARY_API_KEY}" --wait
-    sha256sum ${ARTIFACTS_TO_PUBLISH//checksums.txt/} > checksums.txt
-    echo "🧮     Post Notarization Checksum:"
-    cat checksums.txt
+    local macos_artifacts
+    declare -a macos_artifacts=()
+    while read -r file; do
+      if echo "$file" | grep -q "darwin"; then
+        macos_artifacts+=("${file}")
+        rcodesign sign "${file}" --p12-file="${APPLE_CODESIGN_KEY}" \
+          --code-signature-flags=runtime \
+          --p12-password-file="${APPLE_CODESIGN_PASSWORD_FILE}"
```

Comment:
I don't know to what you are referring exactly. Is it https://www.shellcheck.net/wiki/SC2013 ?

It shows couple valid options on how to iterate lines from a file/pipe/stdin. All of them are equally valid.

The problem is people keep using the invalid iteration patterns. Habits from other programming languages often makes writing shell code harder, as bash is unique in many ways.

Reply:
I meant that this format:

```
grep -v '^ *#' < file | while IFS= read -r line
do
  echo "Line: $line"
done
```

Seems clearer to me, because the set of things you're looping over is at the start.

The linked docs suggest that the form using process substitution (`<(...)`) "avoids a subshell", but I'm not sure I understand why that would be true.  In any case, it feels like having the set of things to loop over at the beginning of the loop is a bit more natural.

Reply:
Both forms are equally valid. Matter of preference, I guess.

Giving things to iterate over before the loop, can be very cloudy, especially if the command is long.

Passing it to `done` is clear to me, and avoids long lines. Feeding data to commands with `<` is quite common in shell. No point, IMHO, of making bash to look like other programming languages, it isn't similar.

Your proposal has only one minor downside. Because it uses pipe, it may fail silently without stopping the script (if `pipefail` option isn't set ofc).